New post about IOMMU (#356)
3mdeb/news-and-ideas: blog/content/post/2020-12-08-IOMMU.md

---
title: What is IOMMU and how it can be used?

> abstract: 'Welcome to a new blogpost series dedicated IOMMU.

Reviewer: ```suggestion
abstract: 'Welcome to a new blogpost series dedicated to IOMMU.
```

          In this article, you can read what IOMMU is
          and find out if its use may be beneficial for you.'
cover: /covers/IOMMU.png
author: marek.kasiewicz
layout: post
published: true
date: 2020-12-30
archives: "2020"

tags:
  - Firmware
  - IOMMU
  - Security
categories:
  - Firmware
  - Security

---

> ## Introduction

Reviewer: ```suggestion
## Introduction

```

isolation and protection by enabling system software to control which areas of
physical memory an I/O device may access. It also helps filter and remap interrupts
from peripheral devices. Let's have a look at IOMMU advantages, disadvantages,
and how it is implemented from the perspective of hardware and software.

Advantages of IOMMU usage:

  - One single contiguous virtual memory region can be mapped to multiple non-contiguous
  physical memory regions. IOMMU can make a non-contiguous memory region appear
  contiguous to a device (scatter/gather). Scatter/gather optimizes streaming DMA
  performance for the I/O device.
  - Memory isolation and protection: device can only access memory regions that are
  mapped for it. Hence faulty and/or malicious devices can't corrupt memory.
  - Memory isolation allows safe device assignment to a virtual machine without
  compromising host and other guest OSes.
  - IOMMU enables 32-bit DMA capable devices to access to > 4GB memory.
  - Support hardware interrupt remapping. It extends limited hardware interrupts
  to software interrupts. Primary uses are the interrupt isolation and translation

>   between interrupt domains (E.g. ioapic vs x2apic on x86)

Reviewer: ```suggestion
  between interrupt domains (e.g. IOAPIC vs x2APIC on x86)
```

Disadvantages of IOMMU usage:

> 

Reviewer: Do we need that empty line?

  - Latency in dynamic DMA mapping, translation overhead penalty.

>   - Host software has to maintain in-memory data structures for use by the IOMMU

Reviewer: Do we need indentation here?

## In hardware

IOMMU’s architecture is designed to accommodate a variety of system topologies.
There can be multiple IOMMUs located at a variety of places in the system fabric.
It can be placed at bridges inside one bus or at bridges between busses of the
same or different types. In modern systems, IOMMU is commonly integrated with the
PCIe root complex. It connects to the PCIe bus from downstream and to the system

> bus ( E.g. infinity fabric) from upstream. One example topology is shown below.

Reviewer: ```suggestion
bus (e.g. infinity fabric) from upstream. One example topology is shown below.
```

![Example Platform Architecture](/img/iommu.png) 

## In software

The IOMMU is configured and controlled via two sets of registers. One in the PCI
configuration space and another set mapped in system address space. Since the
IOMMU appears to OS as a PCI function, it has a capability block in the PCI
configuration space. Additionally, up to eight data structures are placed in the
main system memory. These data structures contain information about the mapping
of interrupts and memory for the usage of peripheral devices. Information about
the device's memory access permissions is stored there too.

## Enabling

IOMMU is a generic name for, be Intel VT-d and AMD-Vi technologies. Make sure

> that your CPU supports one of these before you try to enable IOMMU.

Reviewer: @Sakoram maybe it is worth to mention other implementations, especially IBM one since we entering the OpenPOWER ecosystem and ARM since we have a dedicated team working with that architecture.

Author: I added mentions about other implementation, but I couldn't find how to enable them in the kernel. I can't see a kernel option to enable it on PowerPC and ARM. [docs](https://www.kernel.org/doc/Documentation/admin-guide/kernel-parameters.txt ) say that iommu=on is meant only for x86.

> ### BIOS

Reviewer: ```suggestion
### UEFI/BIOS
```

> First of all, IOMMU has to be initiated by BIOS and information about it has to be

Reviewer: ```suggestion
First of all, IOMMU has to be initiated by UEFI/BIOS, and information about it has to be
```

passed to the kernel in ACPI tables. For the end-user, that means that you have

> to enter the BIOS settings and set IOMMU option to `enabled`. For readers

Reviewer: ```suggestion
to enter the UEFI/BIOS settings and set the IOMMU option to `enabled`. For readers
```

interested in developing firmware capable of IOMMU initialization, the next post

> in this topic will describe the process of enabling IOMMU for APU2 in coreboot.

Reviewer: ```suggestion
in this topic will describe the process of enabling IOMMU for PC Engines apu2 in coreboot.
```

> ### Linux kernel

Reviewer: ```suggestion
### Linux kernel

```

Enable IOMMU support by setting the correct kernel parameter depending on the
type of CPU in use:

- For Intel CPUs (VT-d) set `intel_iommu=on`
- For AMD CPUs (AMD-Vi) set `amd_iommu=on`
- Additionally if you interested in PCIe passthrough set `iommu=pt`

After rebooting, you can check dmesg output to confirm that IOMMU is enabled
```sh

> dmesg | grep -i -e DMAR -e IOMMU

Reviewer: @Sakoram this is valid only for Intel, either we should underline that fact either show also AMD side.

---
title: What is IOMMU and how it can be used?

> abstract: 'Introductory blog post on how IOMMU works.

Reviewer 2: Welcome to a new blogpost series dedicated IOMMU.

>           In this article you can read what IOMMU is

Reviewer 2: ```suggestion
          In this article, you can read what IOMMU is
```

          and find out if its use may be beneficial for you.'
cover: /covers/IOMMU.png
author: marek.kasiewicz
layout: post
published: true
date: 2020-12-08
archives: "2020"

tags:
  - Firmware
  - IOMMU
  - Security
categories:
  - Firmware
  - Security

---

## Introduction

The I/O memory management unit (IOMMU) is a type of memory management unit (MMU)
that connects a Direct Memory Access (DMA) capable expansion bus to the main memory.
It extends the system architecture by adding support for the virtualization of
memory addresses used by peripheral devices. Additionally, it provides memory
isolation and protection by enabling system software to control which areas of
physical memory an I/O device may access. It also helps filter and remap interrupts

> from peripheral devices. 

Reviewer 2: Let's have a look at IOMMU advantages, disadvantages, and purpose in both, hardware and software.

Advantages of IOMMU usage:

  - One single contiguous virtual memory region can be mapped to multiple non-contiguous
  physical memory regions. IOMMU can make a non-contiguous memory region appear
  contiguous to a device (scatter/gather). Scatter/gather optimizes streaming DMA
  performance for the I/O device.
  - Memory isolation and protection: device can only access memory regions that are
  mapped for it. Hence faulty and/or malicious devices can't corrupt memory.
  - Memory isolation allows safe device assignment to a virtual machine without
  compromising host and other guest OSes.
  - IOMMU enables 32-bit DMA capable devices to access to > 4GB memory.
  - Support hardware interrupt remapping. It extends limited hardware interrupts
  to software interrupts. Primary uses are the interrupt isolation and translation
  between interrupt domains (E.g. ioapic vs x2apic on x86)

> Disadvantages:

Reviewer 2: ```suggestion
Disadvantages of IOMMU usage:
```

  - Latency in dynamic DMA mapping, translation overhead penalty.

>   - Host software must maintain in-memory data structures for use by the IOMMU

Reviewer 2: ```suggestion
  - Host software must be maintained in-memory data structures for use by the IOMMU
```

Author: Software is a thing that has to maintain structures, not a thing that has to be maintained.

## In hardware

IOMMU’s architecture is designed to accommodate a variety of system topologies.
There can be multiple IOMMUs located at a variety of places in the system fabric.
It can be placed at bridges inside one bus or at bridges between busses of the

> same or different types.In modern systems, it is commonly integrated with the

Reviewer 2: ```suggestion
same or different types.In modern systems, IOMMU is commonly integrated with the
```

PCIe root complex. It connects to the PCIe bus from downstream and to the system
bus ( E.g. infinity fabric) from upstream. One example topology is shown below.

![Example Platform Architecture](/img/iommu.png) 

## In software

The IOMMU is configured and controlled via two sets of registers. One in the PCI
configuration space and another set mapped in system address space. Since the
IOMMU appears to OS as a PCI function, it has a capability block in the PCI
configuration space. Additionally, up to eight data structures are placed in the
main system memory. These data structures contain information about the mapping
of interrupts and memory for the usage of peripheral devices. Information about
the device's memory access permissions is stored there too.

## Summary

> IOMMU is a useful device with many advantages. Among other things, It protects a

Reviewer 2: It is a very well written piece of text. What  I miss is the length and more of a practical part: Do we use IOMMU? Is it possible to extend this blogpost with some practical, low-effort for you HoWTo part??

Author: I'm not sure if I know any practical way that users can interact with IOMMU. As far as I understand, For 99.9% of users, it is quite an autonomous device. Unless someone writes parts of the kernel/BIOS which manage IOMMU, or writes device drivers this someone won't have to care about anything related to IOMMU. It protects PCs from e.g. DMA attacks, but users don't have to do more than turn it on to get this protection.
The only thing that I can think of is the user's ability to set up IOMMU groups. This is important e.g. while setting up VMs, but the separate post will be dedicated to IOMMU groups.
@miczyg1 Could you correct me if I'm wrong, please?

Reviewer: @Sakoram I recommend the whole post: https://zirblazer.github.io/htmlfiles/coreboot.html#chapter-4.7
More to that I presented AMD IOMMU enabling topic on OSFC 2018: https://3mdeb.com/events/#Open-Source-Firmware-Conference
Finally, we should not show things from an end-user perspective only, our Independent Open Source Firmware Vendor perspective is also important. How complex and important is the enabling of those features? Are we able/capable in supporting other companies (OEM/ODM) in fixing their configuration? What about DMA attacks can we support validation?

Reviewer: More:
https://looking-glass.io/
https://www.reddit.com/r/VFIO/
https://www.kernel.org/doc/Documentation/vfio.txt

Topic is extremely important.

Author: I added a brief description of how to enable IOMMU and what is PCIe Passthrough. If I understand correctly this should be a short introduction to IOMMU (I was told it should have around 600 words, it is ~900, now). The idea is that most of the more complex topic will be presented in future posts.

Reviewer: @Sakoram I like the improvements, but please refer to some external materials, so beginners can improve their knowledge if they want.

Author: @pietrushnic I added a "Further reading" section. Did you read the whole post? Confirmation that everything is accurate from a technical point of view is needed.

Reviewer: @Sakoram yes I did read the post and added a bunch of fixes/suggestions. The post seems to be accurate from a technical point of view. Of course, the topic is huge and in the future it would be great to explore it further.
